sections/3-architecture.md

```diff
@@ -366,3 +366,107 @@ primitives (e.g. ints, strings, etc.), so usage of named
 parameters makes the code more resistant to change and less error-prone,
 versus relying on positioning. The style of indentation chosen here makes
 the instantiation look like a `Map` or a JSON object if you want.
+
+### 3.6 SHOULD put type-class instances as close to the call site as possible
+
+Type-classes are a powerful feature close to functional programming mindset. It 
+favours composition and is central to core `Scala` libraries like 
+[`circe`](https://circe.github.io/circe/), [`cats`](https://typelevel.org/cats/) 
+or even [`scalacheck`](https://www.scalacheck.org/).
+
+One may keep wondering where to put the type-class instances. Each type-class
+has a purpose. Some may be more general like purely functional abstractions 
+from `cats`, some more specific like *json* serialization from `circe`. A good 
+practice is to *put them as close to the call site as possible*. For example:
+
+```scala
+// /core/model/ProductDetail.scala
+
+final case class ProductDetail(id: Int, label: String, description: String)
+
+object ProductDetail {
+  implicit val productDetailEncoder= deriveEncoder
+}
+```
+
+call site:
+
+```scala
+// /access/rest/product/detail/Routes.scala
+
+object Routes {
+  def getProductDetailRoute(id: Id): IO[Response] = {
+    Response.OK(getProductDetail(id).asJson)
+  }
+}
+```
+
+In the code above `ProductDetail` companion object has *json* `Decoder` even though 
+its only use is REST API serialization. Since *json* serialization is specific only
+to REST API all `Encoders` and `Decoders` should be moved as close to this site
+as possible into a `JsonSupport` object. Same example with instances as close to 
+call site as possible:
+
+```scala
+// /core/model/ProductDetail.scala
+
+final case class ProductDetail(id: Int, label: String, description: String)
+```
+
+call site:
+
+```scala
+// /access/rest/product/detail/JsonSupport.scala
+
+object JsonSupport {
+  implicit val productDetailEncoder= deriveEncoder
+}
+```
+
+```scala
+// /access/rest/product/detail/Routes.scala
+
+import io.circe.syntax._
+import JsonSupport.productDetailEncoder
```

Comment:
This is a super simple example. In a real word with a deeply nested structure you need to import all

Reply:
It depends on the project structure. In current sangria code, we actually do not import all of them, since we do not need them. A lot of unused `encoders` and `decoders` were actually removed thanks to treating them explicitly. 

Yes, when you have object/trait with only the necessary set of encoders and decoders that is being used only in 1 place its OK to import them all.